### PR TITLE
v2: typescript declarations

### LINKED ISSUE #403: Include TypeScript Declaration Files
## Expected Behavior

OpenJSCAD.org includes updated TypeScript Declaration Files.

VS Code IntelliSence for **JavaScript** uses **TypeScript** Declation Files.

https://code.visualstudio.com/docs/languages/javascript#_intellisense

DefinetelyTyped has TypeScript Declaration Files for OpenJSCAD.org, but they look a bit old.

https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/openjscad

## Actual Behavior

OpenJSCAD.org does not include TypeScript Declaration Files.

## Steps to Reproduce the Problem

... snip ...

## Specifications

... snip ...

### PR DESCRIPTION
**Pull request by ahdinosaur**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission pass tests?

---

hallo again! :smiley_cat: 

here's a competing pull request to https://github.com/jscad/OpenJSCAD.org/pull/724, will also close #403.

TODO:

- [x] colors
- [x] ~~connectors~~
- [x] curves
  - [x] bezier
- [x] geometries
  - [x] geom2
  - [x] geom3
  - [x] path2
  - [x] poly2
  - [x] poly3
- [x] maths
  - [x] line2
  - [x] line3
  - [x] mat4
  - [x] plane
  - [x] utils
  - [x] vec2
  - [x] vec3
  - [x] vec4
- [x] measurements
- [x] operations
  - [x] booleans
  - [x] expansions
  - [x] extrusions
    - [x] slice
  - [x] hulls
  - [x] transforms
- [x] primitives
- [x] text
- [x] utils

> this demonstrates how we might manually write TypeScript declarations, to get a feel for a TypeScript-fluent approach to types in comparison with using jsdoc to generate types.
>
> personally, i prefer this approach, as even though it's more verbose, the resulting type declarations are much cleaner and there's no awkward tension with jsdoc.
>
> anyways, keen for feedback by existing maintainers, to make sure i'm doing the right approach for this project, before i go ahead. :purple_heart:

### CODE REVIEW

**Comment by z3dev** [comment]
WOW! That's almost another implementation... sigh. Lots of work.

Is there anyway to generate those declarations initially from the JSDOC?

**Comment by ahdinosaur** [comment]
> Is there anyway to generate those declarations initially from the JSDOC?

yeah, that's what #724 does (when you run the `typescript` script), so i could use that to generate the initial declarations (which have mostly `any`s as the types), then fill in the `any`s with actual types.

> Lots of work.

while the types might be tedious af, they're much less work than the actual implementation code that y'all have done! :blush:

**Comment by danmarshall** [comment]
Hi @ahdinosaur, as @z3dev has said I'm a fan of TypeScript, and I've converted a few projects to TypeScript and created a few declarations. I'm happy to see your enthusiasm, and would like to give some pointers which may save a lot of work. I might not personally have much personal time to create the declarations, but perhaps I can help by showing how I've done it in the past.
Hopefully my method is still relevant, I haven't looked at all the latest compiler options. 

My approach is to rename .js files to .ts files and then compile, with the goal of having the compiler's outputted js files being functionally exact as the js source files. That way I know the declarations are compatible. This usually takes some tweaking of the tsconfig.json file. Additionally, I turn on the declarations flag in the tsconfig so I get declarations too. I feel like it's easier than writing them by hand, and hopefully more accurate. This might also seem like a lot of work but I think the exchange ratio of your time to quantity of declaration output is better than the 1:1 of time to write manually.

So, fiddling with the source code in the v2 branch, I've come up with this tsconfig.json:
```
{
    "compilerOptions": {
        "esModuleInterop": true,
        "outDir": "dist",
        "moduleResolution": "node",
        "module": "CommonJS",
        "declaration": true
    },
    "include": ["src"]
}
```
Start by renaming the root `index.js` to `index.ts` and then compile. You should see a `dist/index.d.ts` file which should be blank. Then iterate the following steps, compile again, and watch the file grow with declarations.

* change `module.exports =` to `export =`
* change `... require(...)` to `import ... = require(...)`
* sometimes you'll need to create a temp variable if the `require` is inlined:
  js
  ```
  module.exports = {
    colors: require('./colors'),
  ``` 
  becomes this ts
  ```
  import colors = require('./colors');
  export = {
    colors,
  ```
If this seems crazy, maybe it is. I probably get a weird satisfaction out of seeing the compiled code run. This process also catches other bugs in the codebase, and gives you a lot of insight into the internal parts of the code if you find that interesting. YMMV.

Also, in a monorepo like this, you'd add a tsconfig.json almost everywhere you see a package.json, and apply this one package at a time.

**Comment by ahdinosaur** [comment]
thanks @danmarshall :smiley_cat: 

>     * change `module.exports =` to `export =`
> 
>     * change `... require(...)` to `import ... = require(...)`

oh interesting! i didn't even realize that syntax was possible. should i not use the ES modules syntax?

i was expecting to write declarations such as:

```ts
import Vec2 from '../../maths/vec2/type'
import Mat4 from '../../maths/mat4/type'

export interface Geom2 {
  sides: Array<[Vec2, Vec2]>
  transforms: Mat4
}

export default Geom2
```

or `index.d.ts` and such as:

```ts
export * as colors from './colors'
// ...
```

**Comment by ahdinosaur** [comment]
okay i changed to using the `export =` and `import ... = require(...)`, but i'm getting errors with this change:

> 
>     * sometimes you'll need to create a temp variable if the `require` is inlined:
>       js
>       ```
>       module.exports = {
>         colors: require('./colors'),
>       ```
>       
>       
>       becomes this ts
>       ```
>       import colors = require('./colors');
>       export = {
>         colors,
>       ```

looking around, seems there's conflicting stories about how to do import / exports, some things (like the DefinitelyTyped README) say to [do `export =` and `import ... = require(...)` if doing types for a CommonJS module](https://github.com/DefinitelyTyped/DefinitelyTyped#should-i-add-an-empty-namespace-to-a-package-that-doesnt-export-a-module-to-use-es6-style-imports), but then the TypeScript website [recommends to use ES modules](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html).

i'm tempted to just do the ES module way, even if it means that consumers will need to use `--allowSyntheticDefaultImports` or `--esModuleInterop` compiler options (although i'm not even sure if this would be necessary?). also if #715 happens, i don't want to do this again, as much as i thank CommonJS for enticing me to Node.js for being a simple module system that made sense.

**Comment by z3dev** [comment]
@ahdinosaur Thanks for the supper effort.

I think that you can limit the TS declarations to just the modeling package for now. Hopefully, there will be some early adopters to test out the new declarations and provide some feedback.

**Comment by danmarshall** [comment]
@ahdinosaur - I'm not sure that you can do it the ES module way, since the code seems to be CommonJs. Have you tested it with a TypeScript project, and does it bundle and run? If it works that would be cool, since I also like the syntax better. I would be hesitant to require users to use any particular compiler options since this may not always be compatible when adding in other libraries and may pose a conflict. If #715 happens that would be a major breaking change anyways.

**Comment by ahdinosaur** [comment]
> Have you tested it with a TypeScript project, and does it bundle and run

@danmarshall okay, did a test now, using a fresh directory, `npm init -y`, `tsc --init`, and [`pkl`](https://www.npmjs.com/package/pkl) (not sure how to `npm link` to a lerna module, but this works okay).

seems to work great, the default TypeScript config has `esModuleInterop: true`, which works when using ES modules _or_ the `import ... = require(...)` way. if `esModuleInterop` is set to false, you have to use the `import ... = require(...)` way, but that works just as well. so TypeScript doesn't seem to mind that the declarations are in ES modules while the code is in CommonJS.

this is exciting to see my type declarations show up in my vim to help me autocomplete and show type errors. :relaxed:

**Comment by z3dev** [comment]
@ahdinosaur just a quick question... it seems that the .ts files are replicating the documentation for the functions. is this required?

**Comment by ahdinosaur** [comment]
> @ahdinosaur just a quick question... it seems that the .ts files are replicating the documentation for the functions. is this required?

i think those would be the `.d.ts` files i auto-generated with `tsc` from the current `.js` files, i'm currently manually going through them, fixing them, giving the functions actual types, and removing anything extraneous such as the duplicate documentation. you can track my progress in the [OP](https://github.com/jscad/OpenJSCAD.org/pull/726#issue-512803382), i'm pottering away like the rain here in Wellington. :teddy_bear:

**Comment by z3dev** [comment]
Wellington must be pouring rain.... 

@ahdinosaur Are you hitting any issues?

**Comment by ahdinosaur** [comment]
> @ahdinosaur Are you hitting any issues?

@z3dev, no sorry, got a wee bit distracted by the US election and other things. :smiley_cat: 

will try another push this week, keen to get this over the line. hoping to get things "good enough" that i can start using the types, even if it's not perfect.

**Comment by ahdinosaur** [comment]
@z3dev, one question: for the functions that accept geometries as an argument, do you think it makes more sense to say that argument can be any geometry type (geom2, geom3, poly2, poly3, path2), or do you think it makes more sense to be specific about which geometry types are actually supported by the implementation? said another way, do you expect to eventually support all geometry types in these functions, or are some functions specific to certain geometry types? one random example is `measureArea`.

**Comment by z3dev** [comment]
> @z3dev, one question: for the functions that accept geometries as an argument, do you think it makes more sense to say that argument can be any geometry type (geom2, geom3, poly2, poly3, path2), or do you think it makes more sense to be specific about which geometry types are actually supported by the implementation? said another way, do you expect to eventually support all geometry types in these functions, or are some functions specific to certain geometry types? one random example is `measureArea`.

Correct. All geometries should be supported as arguments to the various operations. There are only a few exceptions now, and should be fixed if failures occur. I think measureArea() behaves fine. Probably offset() will fail if passed a geom3.

**Comment by ahdinosaur** [comment]
hi @z3dev et al, okay this is ready for an initial review. :smiley_cat: 

i've also published with [`gitpkg`](https://github.com/ramasilveyra/gitpkg), so can be installed to test with:

```shell
npm install 'ahdinosaur/OpenJSCAD.org#jscad-modeling-v2.0.0-alpha.6-gitpkg'
```

i only just tested the most basic cases, which seem to work fine:

```js
import { maths } from '@jscad/modeling'

const { vec2 } = maths

console.log(vec2.fromValues(1, 2))
```

```js
import vec2 from '@jscad/modeling/src/maths/vec2'

console.log(vec2.fromValues(1, 2))
```

```js
import vec2 = require('@jscad/modeling/src/maths/vec2')

console.log(vec2.fromValues(1, 2))
```

later will try and port my own project code to use the new types, if that works then i'm happy. :sunny: :seedling:

**Comment by z3dev** [comment]
> hi @z3dev et al, okay this is ready for an initial review.

Super. I added @danmarshall as a reviewer as he will probably use those TS definitions immediately.

I’ll also review.

**Review by z3dev — CHANGES_REQUESTED** [review]
I have a few questions about the management of declarations and tests.

Also, there are a few small requests for changes.

File: packages/modeling/src/geometries/geom2/create.test-d.ts
Comment (ahdinosaur) [reply packages/modeling/src/geometries/geom2/create.test-d.ts:1]: these tests aren't strictly necessary, was mostly a sanity check for myself.

File: packages/modeling/package.json
Comment (ahdinosaur) [reply packages/modeling/package.json:16]: seems like `tsd` only allows this configuration in the `package.json`. but [the configuration](https://github.com/SamVerschueren/tsd/#test-directory) is only to allow tests within the `src` directory, by default it would only check for an `index.test-d.ts` next to the `index.d.ts`, which seems easier to maintain if we want type tests at all.

File: packages/modeling/src/colors/cssColors.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/colors/cssColors.d.ts:149]: as far as i understand these exports _become_ the `cssColors` object. but yeah we could instead define this as an interface (object) where the set of keys are the names and the value is `RGB`. this was more how `tsc` auto-generated it (with `any` types) so i just fixed it (with `RGB` types).

Comment (ahdinosaur) [reply packages/modeling/src/colors/cssColors.d.ts:149]: although, yeah might want to test this works as expected.

File: packages/modeling/src/colors/colorize.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/colors/colorize.d.ts:6]: nah the temporary type is not required, would you prefer i do it that way?

File: packages/modeling/src/colors/cssColors.d.ts
Comment (z3dev) [reply packages/modeling/src/colors/cssColors.d.ts:149]: Either way is fine as long as TS can use these definitions.

File: packages/modeling/src/geometries/geom2/type.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/geometries/geom2/type.d.ts:1]: sure, but i think it's best to split up the definitions in multiple files for the same reasons it's best not to have all your code in a single file. as a specific reason, with the definitions being 1-to-1 with their respective code files, it means you can import at any level and the types will work, whether you import `@jscad/modeling` or `@jscad/modeling/src/maths` it will work. this is better for me because i've been importing `@jscad/modeling/src/maths` and don't want to import things i won't use (and unnecessarily increase bundle size, at least until i've tested this and seen if tree-shaking might work).

File: packages/modeling/src/geometries/path2/appendBezier.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/geometries/path2/appendBezier.d.ts:8]: :+1:

File: packages/modeling/src/colors/colorize.d.ts
Comment (z3dev) [reply packages/modeling/src/colors/colorize.d.ts:6]: Sure. We love clean code. 😃

File: packages/modeling/src/index.test-d.ts
Comment (ahdinosaur) [reply packages/modeling/src/index.test-d.ts:3]: yes, great idea! :+1:

File: packages/modeling/src/operations/booleans/trees/PolygonTreeNode.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/operations/booleans/trees/PolygonTreeNode.d.ts:1]: oops, leftovers from the auto-generating, will remove.

File: packages/modeling/src/geometries/geom2/type.d.ts
Comment (z3dev) [reply packages/modeling/src/geometries/geom2/type.d.ts:1]: Sounds like a good reason to keep the current organization of files.

File: packages/modeling/src/maths/line2/fromPoints.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/maths/line2/fromPoints.d.ts:6]: :+1:

File: packages/modeling/src/maths/line3/fromPoints.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/maths/line3/fromPoints.d.ts:6]: :+1:

File: packages/modeling/src/operations/expansions/expand.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/operations/expansions/expand.d.ts:4]: well, it's a local type (kinda like a local var) of `Geometry` without `Poly3`, unless the `Geometry` i define shouldn't have `Poly3` in the first place?

File: packages/modeling/src/operations/expansions/expand.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/operations/expansions/expand.d.ts:14]: :+1:

File: packages/modeling/src/operations/expansions/offset.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/operations/expansions/offset.d.ts:14]: :+1:

File: packages/modeling/src/operations/expansions/expand.d.ts
Comment (z3dev) [reply packages/modeling/src/operations/expansions/expand.d.ts:4]: Ok. Understood.

File: packages/modeling/src/operations/extrusions/extrudeFromSlices.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/operations/extrusions/extrudeFromSlices.d.ts:8]: this is a "generic type", which you can think of as a type that takes a type as an argument (similar to how a function takes a value or reference as an argument). in this case, `Base` is a generic type for the `base` variable used in the callback. it's generic because it's up to the caller to decide what the base type is, or at least that's my understanding, since it doesn't seem to matter what you pass in as long as eventually the callback returns slices.

i will say, `extrudeFromSlices` is my new favorite learning about jscad, which i learned a month ago or so. :smiley_cat:

File: packages/modeling/src/operations/transforms/scale.d.ts
Comment (ahdinosaur) [reply packages/modeling/src/operations/transforms/scale.d.ts]: :+1:

**Review by z3dev — DISMISSED** [review]
OK. All my little nits have been corrected. Thanks @ahdinosaur

I'll wait for @danmarshall to either complete a quick review before merging.

**Comment by ahdinosaur** [comment]
i rebased the commits, to rename the messages. i was originally planning to squash all into one commit, but as far as i understand y'all would prefer individual commits?

**Comment by ahdinosaur** [comment]
published the latest using `gitpkg`, can be installed with:

```shell
npm install 'ahdinosaur/OpenJSCAD.org#jscad-modeling-v2.0.0-alpha.7-gitpkg'
```

**Comment by z3dev** [comment]
> i rebased the commits, to rename the messages. i was originally planning to squash all into one commit, but as far as i understand y'all would prefer individual commits?

The final merge will be a squashed merge, so no need to squash yourself. The comments on the commits are very useful during reviews. 👍

**Comment by z3dev** [comment]
@danmarshall can you review? thoughts?

i'm ready to merge this but wanted another review / opinion.

**Comment by ahdinosaur** [comment]
started porting over a project's code to use the types, most of it works really well, some of it needs some iron-ing out. :smiley_cat: 

the lower-level math and geometry types seems good, will need to think about the higher-level operations that can apply to arbitrary geometries, seems there's mismatches sometimes trying to pass around generic geometries, some functions get confused when they expect specific geometries.

still keen to merge this pull request, if the overall approach is sound, and i'll be fixing things up as i (and hopefully others) put this to more use.

 :sunflower:

**Comment by danmarshall** [comment]
Sorry, I haven't had much time to look at this... Hoping someone else in the community can test it out.

File: packages/modeling/src/index.test-d.ts
Comment (z3dev) [reply packages/modeling/src/index.test-d.ts:3]: @ahdinosaur please put this on your TODO list for the next set of declaration changes. ;)

**Review by z3dev — APPROVED** [review]
Approved. Let's get these released and used. Hopefully, there will be some good feedback.

**Comment by danmarshall** [comment]
Hi @ahdinosaur - I may finally have some time this weekend to try these out :)
Looks like these are part of @jscad/modeling@2.0.0-alpha.7, but when I install I get @jscad/modeling@2.0.0-alpha.4
@z3dev will you be publishing @jscad/modeling@2.0.0-alpha.7 soon?

**Comment by z3dev** [comment]
@danmarshall excellent!

I took a look at NPM. I have no idea but here’s what NPM says about the versions.

Version               Downloads (Last 7 Days)   Published
2.0.0-alpha.10   29                                         alpha
2.0.0-alpha.4     30                                         latest

For some reason, NPM is stuck at alpha.4. Did I mention that NPM sucks?

If you use the ‘alpha’ tag then it will install the latest version; alpha.10.

I’ll have to research this further...

**Comment by kenianbei** [comment]
Just wanted to say I've been using the modeling types and haven't come across anything wrong yet. I'd also like to get types done for the regl-renderer, which should be less complex since regl is typed. Right now I'm pulling the types directly from regl. I may work on that next week if I have time.

**Comment by z3dev** [comment]
> Just wanted to say I've been using the modeling types and haven't come across anything wrong yet. I'd also like to get types done for the regl-renderer, which should be less complex since regl is typed. Right now I'm pulling the types directly from regl. I may work on that next week if I have time.

That would be cool. Also, there’s not much documentation for the real-renderer yet so feel free to complete some of the documentation for those functions that need to be exposed.

**Pull request merged**
